graylog2-server/src/main/java/org/graylog2/inputs/transports/KafkaTransport.java

```diff
@@ -173,17 +187,124 @@ public void setMessageAggregator(CodecAggregator ignored) {
+    }
+
+    private void doLaunchConsumer(final MessageInput input) {
+        final Properties props = new Properties();
+
+        props.put("group.id", configuration.getString(CK_GROUP_ID, DEFAULT_GROUP_ID));
+        props.put("fetch.min.bytes", String.valueOf(configuration.getInt(CK_FETCH_MIN_BYTES)));
+        props.put("fetch.max.wait.ms", String.valueOf(configuration.getInt(CK_FETCH_WAIT_MAX)));
+        //noinspection ConstantConditions
+        props.put("bootstrap.servers", configuration.getString(CK_BOOTSTRAP));
+        // Map largest -> latest, smallest -> earliest
+        final String resetValue = configuration.getString(CK_OFFSET_RESET, DEFAULT_OFFSET_RESET);
+        props.put("auto.offset.reset", resetValue.equals("largest") ? "latest" : "earliest");
+        // Default auto commit interval is 60 seconds. Reduce to 1 second to minimize message duplication
+        // if something breaks.
+        props.put("auto.commit.interval.ms", "1000");
+        props.put(org.apache.kafka.clients.consumer.ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        props.put(org.apache.kafka.clients.consumer.ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+
+        insertCustomProperties(props);
+
+        final int numThreads = configuration.getInt(CK_THREADS);
+        final ExecutorService executor = executorService(numThreads);
+        // this is being used during shutdown to first stop all submitted jobs before committing the offsets back to zookeeper
+        // and then shutting down the connection.
+        // this is to avoid yanking away the connection from the consumer runnables
+        stopLatch = new CountDownLatch(numThreads);
+
+        IntStream.range(0, numThreads).forEach(i -> {
+            executor.submit(() -> {
+                final Properties nprops = (Properties) props.clone();
+                nprops.put("client.id", "gl2-" + nodeId + "-" + input.getId() + "-" + i);
+                final KafkaConsumer<byte[], byte[]> consumer;
+                try {
+                    consumer = new KafkaConsumer<>(nprops);
+                    //noinspection ConstantConditions
+                    consumer.subscribe(Pattern.compile(configuration.getString(CK_TOPIC_FILTER)), new NoOpConsumerRebalanceListener());
+                } catch (Exception e) {
+                    LOG.warn("Could not create KafkaConsumer", e);
+                    throw e;
+                }
+
+                while (!stopped) {
+                    final ConsumerRecords<byte[], byte[]> consumerRecords = consumer.poll(1000);
+                    final Iterator<ConsumerRecord<byte[], byte[]>> consumerIterator = consumerRecords.iterator();
+                    try {
+                        // we have to use hasNext() here instead foreach, because next() marks the message as processed immediately
```

Comment:
Is that comment still valid? I checked the code briefly and I cannot see where `next()` would mark the message as processed.

Reply:
I guess that depends. The default is `auto.commit.enable` = true. So I'd say it does.

Reply:
Nevermind. You're right :)